### PR TITLE
feat(stock): algorithm for quantity in stock

### LINKED ISSUE #4864: CMM calculation using days of stock
We are currently calculating the CMM using the days of stock consumption but it has been shown by several formulas that it's better to do it based on the days of stock.

I suggest we can create a table with aggregated data for each inventory and it should help us to switch to whatever formula we want.

you can an example below :

stock_movement_status

Here are formulas that we currently have :
```
Algo1 = (∑consumed_quantity )/(∑𝑆𝑡𝑜𝑐𝑘_𝑑𝑎𝑦)*30.5
```

```
Algo2 = ((∑𝑐𝑜𝑛𝑠𝑢𝑚𝑒𝑑_𝑞𝑢𝑎𝑛𝑡𝑖𝑡𝑦)/(∑𝑐𝑜𝑛𝑠𝑢𝑚𝑝𝑡𝑖𝑜𝑛_𝑑𝑎𝑦))  × 30.5
```

```
Algo3 =((∑consumed_quantity)/(∑𝑑𝑎𝑦𝑠))*30.5
```
```
Algo_msh = ((∑consumed_quantity)/( #month − ((∑stock_out_days)/30.5) ))
```
Let me show you how we can get all the informations we need for calculating the CMM.

Our calculation of the CMM we be applied during the period  ```2019-02-01```  - ```2020-03-31```

This beautiful SQL query below will help us to retrieve data during this our period

```sql


SET @start_date = '2019-02-01';
SET @end_date = '2020-03-31';
SET @inventory_name = 'amoxycilline 250 mg';



SET @last_inventory_mvt_date = NULL;
--
SELECT `end_date` 
INTO  @last_inventory_mvt_date
FROM stock_movement_status m 
JOIN inventory i ON m.inventory_uuid = i.uuid
WHERE i.text = 'amoxycilline 250 mg' 
ORDER BY `end_date` DESC LIMIT 1;


SELECT x.start_date, x.end_date, x.in_quantity, x.out_quantity, x.in_stock_quantity,
	 x.inventory, x.depot, (DATEDIFF(x.end_date, x.start_date) +1) AS frequency
FROM (
SELECT 

  IF(m.start_date <@start_date, @start_date, m.start_date) AS start_date , 
  IF(m.end_date > @end_date, @end_date,  IF(m.end_date = @last_inventory_mvt_date AND 
  		@last_inventory_mvt_date < @end_date, @end_date, m.end_date )) AS end_date,
  
  in_quantity, out_quantity, m.quantity AS `in_stock_quantity`,
  i.text as 'inventory', d.text AS `depot`
FROM stock_movement_status m
JOIN depot d ON d.uuid = m.depot_uuid
JOIN inventory i ON i.uuid = m.inventory_uuid

WHERE i.text = 'amoxycilline 250 mg' 
	AND  ((m.start_date >= @start_date) OR  (m.start_date <= @start_date AND m.end_date >= @start_date)) 
	AND ( (m.end_date <=  @end_date) OR ( m.start_date < @end_date ))
ORDER BY d.text
) AS x

```
The inventory_name will be replace by it ```uuid```  and all those queries will be in a store procedure or a function ,don't worry about that.

You can see the result in the table as follow
extracted data

Let edit these queries and calculate the CMM for each algorithm

```sql


SET @start_date = '2019-02-01';
SET @end_date = '2020-03-31';
SET @inventory_name = 'amoxycilline 250 mg';


SET @last_inventory_mvt_date = NULL;
--
SELECT `end_date` 
INTO  @last_inventory_mvt_date
FROM stock_movement_status m 
JOIN inventory i ON m.inventory_uuid = i.uuid
WHERE i.text = 'amoxycilline 250 mg' 
ORDER BY `end_date` DESC LIMIT 1;


SET @sum_consumed_quantity = 0;
SET @sum_stock_day = 0;
SET @sum_consumption_day =0;
SET @sum_stock_out_days = 0;

SET @sum_days = DATEDIFF( @end_date,  @start_date) + 1;
SET @number_of_month = (DATEDIFF(@end_date,  @start_date) + 1)/30.5;




SELECT 
    SUM(cmm_data.out_quantity) AS sum_consumed_quantity, 
	SUM(IF(cmm_data.out_quantity <> 0, cmm_data.frequency, 0)) AS sum_consumption_day,
	SUM(IF(cmm_data.in_stock_quantity = 0, cmm_data.frequency, 0)) AS sum_stock_out_days,
	SUM(IF(cmm_data.in_stock_quantity <> 0, cmm_data.frequency, 0)) AS sum_stock_day
INTO @sum_consumed_quantity, @sum_consumption_day, @sum_stock_day, @sum_stock_out_days
FROM (
    SELECT x.start_date, x.end_date, x.in_quantity, x.out_quantity, x.in_stock_quantity,
        x.inventory, x.depot, (DATEDIFF(x.end_date, x.start_date) +1) AS frequency
    FROM (
        SELECT 

        IF(m.start_date <@start_date, @start_date, m.start_date) AS start_date , 
        IF(m.end_date > @end_date, @end_date,  IF(m.end_date = @last_inventory_mvt_date AND 
                @last_inventory_mvt_date < @end_date, @end_date, m.end_date )) AS end_date,
        
        in_quantity, out_quantity, m.quantity AS `in_stock_quantity`,
        i.text as 'inventory', d.text AS `depot`
        FROM stock_movement_status m
        JOIN depot d ON d.uuid = m.depot_uuid
        JOIN inventory i ON i.uuid = m.inventory_uuid

        WHERE i.text = 'amoxycilline 250 mg' 
            AND  ((m.start_date >= @start_date) OR  (m.start_date <= @start_date AND m.end_date >= @start_date)) 
            AND ( (m.end_date <=  @end_date) OR ( m.start_date < @end_date ))
        ORDER BY d.text
    ) AS `x`
) AS `cmm_data`;



SET @Algo1 = ( @sum_consumed_quantity/(IF((@sum_stock_day = NULL) OR (@sum_stock_day = 0),1, @sum_stock_day)))*30.5;

SET @Algo2 = ((@sum_consumed_quantity)/(IF( (@sum_consumption_day = NULL) OR @sum_consumption_day = 0,1, @sum_consumption_day)))  * 30.5;

SET @Algo3 =((@sum_consumed_quantity)/ (IF(@sum_days = NULL OR @sum_days = 0,1, @sum_days)))*30.5;

SET @Algo_msh = (@sum_consumed_quantity/(@number_of_month - (@sum_stock_out_days/30.5) ));


SELECT @sum_consumed_quantity, @sum_consumption_day, @sum_stock_day, @sum_stock_out_days, @number_of_month, @sum_days;
SELECT @Algo1, @Algo2, @Algo3, @Algo_msh;
```

The result will be :

variables

result

### PR DESCRIPTION
**Pull request by jniles**
Adds an algorithm developed by @jeremielodi and formatted by @jniles for computing the quantity in stock of inventory items.

Closes #4864.

### CODE REVIEW

File: server/models/admin.sql
Comment (jmcameron) [inline server/models/admin.sql:207]: Aren't we going to need an end date too?   For instance, we may need to compute these things over a month but may be several days into the next month.

File: server/models/procedures/stock.sql
Comment (jmcameron) [inline server/models/procedures/stock.sql]: Don't you want to drop this table in the main routine above?  Or is "temporary table" enough to avoid reusing the table between queries?

File: server/models/procedures/stock.sql
Comment (jmcameron) [inline server/models/procedures/stock.sql]: throw -> through

**Review by jmcameron — APPROVED** [review]
I was impressed by this SQL code.  Nice job @jeremielodi !

I'm pretty sure we will want to be able to specify the end date for these searches.   Someone wanting to do them through the current date can just pass in now().

Could you give some statistics about how long this takes to run on the various databases (please include the number of  moments involved so we get a sense of how this scales).    

I found the labels for the aggregate data a little confusing.  A possible suggestion for making that clearer

- in_quantity => start_quantity
- out_quantity => consumed  (or quantity_consumed)
- in_stock_quantity => end_quantity

(I presume that is what these columns mean.)
Note that these names correspond with the "Start Date" and "End Date"

**Comment by jmcameron** [comment]
It also occurred to me that when we compute CMM using the "Days in Stock", we need to make sure that "Days in stock" is the number of days that the depot was open AND there was stock.   Having stock or not during days that there is no chance to dispense/transfer stock is irrelevant to CMM.  That implies that the 30.5 number we see in the CMM formulas should probably be replaced by the maximum number days the facility is open during the period in question.  Note that this number (max days in stock) depends on the depot and will vary from month to month (depending on weekends hit, holidays, closures, etc).

File: server/models/procedures/stock.sql
Comment (jeremielodi) [reply server/models/procedures/stock.sql]: It's should be a permanent table, I used a temporary table just for the demostration

Comment (jeremielodi) [reply server/models/procedures/stock.sql]: @jniles we don't need to store the ```inventory name``` as we already have it ```uuid```

**Review by jeremielodi — COMMENTED** [review]
Good work @jniles, I wonder why you don't call the procedure ```computeStockQuantity``` after a stock movement registration.

**Review by mbayopanda — APPROVED** [review]
The procedures work without errors, I am wondering just we have there temporary tables so how can we correctly test the procedure ?

**Comment by jeremielodi** [comment]
As we said this table must be like ```period_total``` we should create it in the schema and make it permanent.

File: server/models/procedures/stock.sql
Comment (lomamech) [inline server/models/procedures/stock.sql:291]: If there will not be an end date, the best will be to simply use date instead of start date

File: server/models/procedures/stock.sql
Comment (lomamech) [inline server/models/procedures/stock.sql]: but also since we will use a permanent table instead of a temporary table I find that it is a good proposal

**Review by lomamech — APPROVED** [review]

**Comment by jmcameron** [comment]
Please explain again why these queries do not need an end date.   It seems that it could be useful in may research/analysis scenarios.

**Comment by jniles** [comment]
@jmcameron, the idea is to simply create a persistent table to store the quantity in stock for an inventory in a depot.  Currently it's a temporary table, but as many of my teammates have said, it should be permanent.  I haven't really had the chance to code in a few days so I didn't get around to making those changes.   Hopefully @jeremielodi will.

This isn't actually proposing an API to query the table - that can be done directly in SQL.  Instead, this will be run after stock movements have occurred to update the information in the table. 

The start date exists for performance reasons.  If you pre-date a stock movement (e.g. record the fact that stock came in last month) that will affect all stock balances up until today.  So, rather than recompute for all time, we can just recompute from that date to the present.

Does that make sense?

**Comment by jmcameron** [comment]
> @jmcameron, the idea is to simply create a persistent table to store the quantity in stock for an inventory in a depot. 
> Does that make sense?

Thanks.  That helps.  I recall the discussion about a persistent table of day-to-day totals, but had not quite connected that this PR/code was about that.  Yes, keeping it up to date as stock is dispensed is the right and efficient way to do it and no end date is needed.

**Comment by jmcameron** [comment]
Given what we are doing, I think retaining 'start_date' rather than just 'date' is best.

File: server/controllers/stock/index.js
Comment (mbayopanda) [inline server/controllers/stock/index.js:155]: 🐐

File: server/models/migrations/next/migrate.sql
Comment (mbayopanda) [inline server/models/migrations/next/migrate.sql:740]: 👍

**Review by mbayopanda — COMMENTED** [review]
The stock entry and stock exit still working fine

We are writing successfully in the stock_movement_status

There is just one thing i noticed is `in_quantity` and `out_quantity` are not yet computed in the stock_movement_status table 

image

**Comment by jniles** [comment]
@mbayopanda To allow us to advance rapidly, I think I'll merge this as is (it has many approvals already), and open an issue about the `in_quantity` and `out_quantity`.  They aren't actually necessary for the CMM calculation.

**Review by jniles — COMMENTED** [review]
bors r=mbayopanda,jmcameron,lomamech

**Review by jniles — COMMENTED** [review]
bors r=jmcameron,mbayopanda,lomamech

File: server/models/procedures/stock.sql
Comment (jmcameron) [reply server/models/procedures/stock.sql]: Thanks for the explanations.  That helped me understand what you were doing here.

**Comment by jmcameron** [comment]
Should this still be marked with the DO NOT MERGE label?

File: server/controllers/stock/index.js
Comment (jeremielodi) [reply server/controllers/stock/index.js:155]: I know why this error occur(in_quantity=0 and out_quantity =0), I will try to solve the issue when it will be created

Comment (jeremielodi) [reply server/controllers/stock/index.js:155]: As this PR is not merged yet @jniles can update the procedure ```ComputeStockQuantiy``` as follow 

```sql


DROP PROCEDURE IF EXISTS `computeStockQuantity`$$
CREATE PROCEDURE `computeStockQuantity` (
  IN _start_date DATE,
  IN _inventory_uuid BINARY(16),
  IN _depot_filter_uuid BINARY(16)
) BEGIN
  DECLARE done BOOLEAN;
  DECLARE _depot_uuid, _row_uuid BINARY(16);
  DECLARE _end_date DATE;
  DECLARE _qtt, _in_qtt, _out_qtt DECIMAL(19, 4);




  DROP TEMPORARY TABLE IF EXISTS `temp_stock_movement`;

    -- every stock movement for a inventory will be stored here in order to facilitate search
  CREATE TEMPORARY TABLE `temp_stock_movement` (
    `reference` INT,
    `date` DATE,
    `depot_uuid` BINARY(16),
    `inventory_uuid` BINARY(16),
    `quantity` DECIMAL(19,4),
    `in_quantity` DECIMAL(19,4),
    `out_quantity` DECIMAL(19,4),
    KEY `depot_uuid` (`depot_uuid`),
    KEY `inventory_uuid` (`inventory_uuid`),
    KEY `reference` (`reference`)
  ) ENGINE=InnoDB DEFAULT CHARACTER SET = utf8mb4 DEFAULT COLLATE = utf8mb4_unicode_ci;

  -- every stock movement for a inventory will be stored here in order to facilitate search
  CREATE TEMPORARY TABLE IF NOT EXISTS `temp_depot` (
    `reference` INT,
    `uuid` BINARY(16),
    `text` varchar(100),
    KEY `uuid` (`uuid`),
    KEY `reference` (`reference`)
  ) ENGINE=InnoDB DEFAULT CHARACTER SET = utf8mb4 DEFAULT COLLATE = utf8mb4_unicode_ci;

  SET _end_date = NOW();
  SET _start_date = DATE(_start_date);
  SET @depot_number = 0;

  DELETE FROM `temp_depot`;

  SET @filter_by_depot = (_depot_filter_uuid IS NOT NULL);

  IF @filter_by_depot = 0  THEN -- will compute inventory's quantity in each depot
    INSERT INTO `temp_depot`
    SELECT (@depot_number:=@depot_number + 1)as ref, `uuid`, `text`
    FROM depot;
  ELSE
    INSERT INTO `temp_depot`  -- will compute inventory's quantity just in a specific depot
    SELECT (@depot_number := @depot_number + 1) as ref, `uuid`, `text`
    FROM depot
    WHERE `uuid` = _depot_filter_uuid;
  END IF;

  SET @depot_counter = 1;
  -- let's loop through all inventories

  WHILE (@depot_counter <= @depot_number) DO
    SET @depot_uuid = (SELECT `uuid` FROM `temp_depot` WHERE `reference` = @depot_counter);

    -- delete all inventory's data for this period

    DELETE FROM `stock_movement_status`
    WHERE  `start_date` >= _start_date AND  `end_date` <= _end_date AND  `inventory_uuid` = _inventory_uuid AND `depot_uuid` =  @depot_uuid;

    DELETE FROM `temp_stock_movement` WHERE 1;
    INSERT INTO `temp_stock_movement`

    SELECT (@row_number:=@row_number + 1) AS `reference`, x.* FROM (
      SELECT DATE(m.date) AS `date`,  m.depot_uuid, i.uuid,
        SUM( IFNULL(m.quantity * IF(m.is_exit = 1, -1, 1), 0)) AS quantity,
        SUM(IF(m.is_exit = 0, m.quantity, 0)),
        SUM(IF(m.is_exit = 1, m.quantity, 0))
      FROM `stock_movement` m
        JOIN lot l ON l.uuid = m.lot_uuid
        JOIN inventory i ON i.uuid = l.inventory_uuid
      WHERE i.uuid = _inventory_uuid AND DATE(m.date) <= DATE(_end_date) AND m.depot_uuid =  @depot_uuid
      GROUP BY  DATE(m.date), m.depot_uuid
      ORDER BY  `date`
    ) AS x
    ORDER BY x.date;

	 
    SET @row_i = 0;
    SET _row_uuid = NULL;

    SELECT SUM(quantity) AS quantity INTO _qtt
    FROM temp_stock_movement m
      JOIN inventory i ON i.uuid = m.inventory_uuid
      JOIN depot d ON d.uuid = m.depot_uuid
    WHERE DATE(m.date) <=_start_date AND m.depot_uuid = @depot_uuid;

    SET _qtt = IFNULL(_qtt, 0);

      SELECT in_quantity, out_quantity 
      INTO _in_qtt, _out_qtt
      FROM temp_stock_movement m
      WHERE DATE(m.date) =DATE(_start_date) AND m.depot_uuid = @depot_uuid; 
 
    DELETE FROM temp_stock_movement WHERE DATE(`date`) <=_start_date AND `depot_uuid` = @depot_uuid;

    -- check if this date already exist in stock_movement_status for the inventory
    SET @date_exists = 0;

    SELECT `uuid`, count(`uuid`) as nbr
      INTO _row_uuid,  @date_exists
    FROM `stock_movement_status`
    WHERE `depot_uuid` = @depot_uuid AND `inventory_uuid` = _inventory_uuid
      AND DATE(`start_date`) <= _start_date
    ORDER BY `start_date` DESC
    LIMIT 1;

    IF @date_exists = 0 THEN
      SET _row_uuid  = HUID(uuid());
      INSERT INTO  `stock_movement_status` VALUES (_row_uuid, _start_date, _start_date, _qtt, IFNULL(_in_qtt, 0), IFNULL(_out_qtt, 0), _inventory_uuid, @depot_uuid);
    END IF;

    SELECT `reference` INTO @row_i
    FROM temp_stock_movement
    WHERE `date` > _start_date
    LIMIT 1;

    SET @mvts_number = (SELECT COUNT(*) from temp_stock_movement  WHERE `date` > _start_date);
    SET  @row_number = @row_i + @mvts_number - 1;

    SET @row_i = IFNULL(@row_i, 0);

    IF @row_i = 0 THEN
      UPDATE `stock_movement_status` SET `end_date` = _end_date WHERE `uuid` = _row_uuid;
    ELSE
      WHILE (@row_i < @row_number + 1) DO
        SET _in_qtt = 0;
        SET _out_qtt = 0;

        SET @current_qtt = 0;
        SET @current_date = _start_date;

        SELECT  m.quantity, m.in_quantity, m.out_quantity, m.date
          INTO @current_qtt, _in_qtt, _out_qtt, @current_date
        FROM temp_stock_movement m
        JOIN inventory i ON i.uuid = m.inventory_uuid
        JOIN depot d ON d.uuid = m.depot_uuid
        WHERE m.reference = @row_i
        GROUP BY m.depot_uuid;
        UPDATE `stock_movement_status` SET `end_date` = DATE_SUB(@current_date, INTERVAL 1 DAY) WHERE `uuid` = _row_uuid;
        IF @current_qtt <> _qtt THEN
          SET _row_uuid = HUID(UUID());
          SET _qtt = @current_qtt + _qtt;
          INSERT INTO  `stock_movement_status` VALUES (_row_uuid, @current_date, @current_date, _qtt, _in_qtt, _out_qtt, _inventory_uuid,  @depot_uuid);
        END IF;
        SET @row_i= @row_i + 1;
      END WHILE;
    END IF;

    UPDATE `stock_movement_status` SET `end_date` = _end_date WHERE `uuid` = _row_uuid;
    SET _qtt = 0;
    SET @depot_counter = @depot_counter +1;
    SET _row_uuid = NULL;
  END WHILE;
END$$

```

Before the line where we delete some movement in ```temp_stock_movement ``` that we won't need anymore
I get the in and out quantity of the start_date 
```sql     

      
      SELECT in_quantity, out_quantity 
      INTO _in_qtt, _out_qtt
      FROM temp_stock_movement m
      WHERE DATE(m.date) =DATE(_start_date) AND m.depot_uuid = @depot_uuid; 
 
    DELETE FROM temp_stock_movement WHERE DATE(`date`) <=_start_date AND `depot_uuid` = @depot_uuid;
```
After that I check in the ```stock_movement_status``` if there is a quantity computation of this inventory at that date(start_date), In the issue of @mbayopanda nothing will be found because the record of ```2020-09-01``` was deleted in order to recompute.
You should know that the procedure delete every row in ```stock_movement_status``` for the given period

To finish, I used the _in_qtt and _out_qtt here
```sql
    IF @date_exists = 0 THEN
      SET _row_uuid  = HUID(uuid());
      INSERT INTO  `stock_movement_status` VALUES (_row_uuid, _start_date, _start_date, _qtt, IFNULL(_in_qtt, 0), IFNULL(_out_qtt, 0), _inventory_uuid, @depot_uuid);
    END IF;
```

**Comment by jniles** [comment]
> 
> 
> Should this still be marked with the DO NOT MERGE label?

True.  Sorry!

**Comment by jniles** [comment]
bors r=jmcameron,mbayopanda,lomamech

**Comment by jniles** [comment]
@jeremielodi this has landed.  If you want to make adjustments to it, go for it!

**Comment by jeremielodi** [comment]
okay, I see that added my last correction. thanks

**Comment by jeremielodi** [comment]
Oh! You didn't. Let me create an issue for that then

**Pull request merged**
